src/boot_O2/__osMalloc.c

```diff
@@ -45,181 +49,352 @@ ArenaNode* ArenaImpl_GetLastBlock(Arena* arena) {
+ * the freed pointer is undefined behaviour.
+ *
+ * @param[in, out] arena  The specific Arena to be used for the allocation.
+ * @param[in, out] ptr    The allocated memory block to deallocate.
+ */
 void __osFree(Arena* arena, void* ptr) {
     ArenaNode* node;
     ArenaNode* next;
     ArenaNode* prev;
-    ArenaNode* newNext;
 
     ArenaImpl_Lock(arena);
-    node = (ArenaNode*)((u32)ptr - sizeof(ArenaNode));
 
-    if (ptr == NULL || (node->magic != NODE_MAGIC) || node->isFree) {
-        goto end;
-    }
+    node = (ArenaNode*)((uintptr_t)ptr - sizeof(ArenaNode));
 
-    next = node->next;
-    prev = node->prev;
-    node->isFree = 1;
+    if (ptr != NULL && node->magic == NODE_MAGIC && !node->isFree) {
+        next = node->next;
+        prev = node->prev;
+        node->isFree = true;
+
+        // Checks if the next node is contiguous to the current node and if it isn't currently allocated. Then merge the
+        // two nodes into one.
+        if ((uintptr_t)next == (uintptr_t)node + sizeof(ArenaNode) + node->size && next->isFree) {
+            ArenaNode* newNext = next->next;
+
+            if (newNext != NULL) {
+                newNext->prev = node;
+            }
 
-    newNext = next;
-    if ((u32)next == (u32)node + sizeof(ArenaNode) + node->size && next->isFree) {
-        newNext = next->next;
-        if (newNext != NULL) {
-            newNext->prev = node;
+            node->size += next->size + sizeof(ArenaNode);
+
+            node->next = newNext;
+            next = newNext;
         }
 
-        node->size += next->size + sizeof(ArenaNode);
+        // Checks if the previous node is contiguous to the current node and if it isn't currently allocated. Then merge
+        // the two nodes into one.
+        if (prev != NULL && prev->isFree && (uintptr_t)node == (uintptr_t)prev + sizeof(ArenaNode) + prev->size) {
+            if (next != NULL) {
+                next->prev = prev;
+            }
 
-        node->next = newNext;
-        next = newNext;
+            prev->next = next;
+            prev->size += node->size + sizeof(ArenaNode);
+        }
     }
 
-    if (prev != NULL && prev->isFree && (u32)node == (u32)prev + sizeof(ArenaNode) + prev->size) {
-        if (next) {
-            next->prev = prev;
+    ArenaImpl_Unlock(arena);
+}
+
+/**
+ * @brief Reallocates the pointer \p ptr.
+ * The pointer \p ptr must be either a pointer previously allocated by `__osMalloc`, `__osMallocR` or `__osRealloc` and
+ * not freed yet, or a `NULL` pointer.
+ *
+ * If \p ptr is `NULL` a new pointer is allocated. See `__osMalloc` for more details.
+ * If \p newSize is 0, then the given pointer is freed and `NULL` is returned. See `__osFree` for more details.
+ * If \p newSize is bigger than the currently allocated allocated pointer, then the area of memory is expanded to a size
+ * big enough to fit the requested size.
```

Comment:
This looks like an unordered list.

Reply:
I blame the formater for messing up this